presto-spi/src/main/java/com/facebook/presto/spi/ColumnHandle.java

```diff
@@ -41,4 +41,9 @@ default ColumnHandle withRequiredSubfields(List<Subfield> subfields)
     {
         return this;
     }
+
+    default boolean isPartitionKey()
```

Comment:
partition key is a hive specific concept.  I don't think it's right to add it to the SPI.

Reply:
> partition key is a hive specific concept. I don't think it's right to add it to the SPI.

Actually, it's there in some other connectors as well. And even Oracle has partitioned tables. So this is not a hive-only concept:

https://docs.oracle.com/cd/B10500_01/server.920/a96524/c12parti.htm

Reply:
I think we should discuss in-depth if we want to expose partition related concepts to SPI level as it is not a generic connector concepts. Many sources do not have partition concept.

Regarding this change, I believe we are looking to skip optimization for group by keys having low cardinality. Should we leverage column statistics to make these decisions? At least for partition columns we do have these stats available in the metadata, so it should be possible to leverage it.

Reply:
> I think we should discuss in-depth if we want to expose partition related concepts to SPI level as it is not a generic connector concepts. Many sources do not have partition concept.
> 
> Regarding this change, I believe we are looking to skip optimization for group by keys having low cardinality. Should we leverage column statistics to make these decisions? At least for partition columns we do have these stats available in the metadata, so it should be possible to leverage it.

See my comment above. We have it in a few connectors and ithe concept does exist in many databases. I think making it difficult to access we are not able to leverage these things properly.  This definitely should be here as Presto can do things better for partition keys (like better aggregation planning that I'm working on) so any connector that can support this concept will benefit.

Reply:
> I think we should discuss in-depth if we want to expose partition related concepts to SPI level as it is not a generic connector concepts. Many sources do not have partition concept.
> 
> Regarding this change, I believe we are looking to skip optimization for group by keys having low cardinality. Should we leverage column statistics to make these decisions? At least for partition columns we do have these stats available in the metadata, so it should be possible to leverage it.

As for column statistics - I don't think they are reliable. This PR is mainly setting up for better aggregation optimizations for partition keys that I'm working on.

Reply:
BTW the current singular method (marked experimental) belongs even less there. So looks like this API was added for convenience rather than properly designed spi. So we can may be take a separate action to improve it and make it more useful.

Reply:
> "As for column statistics - I don't think they are reliable"

Can you please share more light on it? We rely on them for fragment result caching and partition pruning.

Reply:
> "This PR is mainly setting up for better aggregation optimizations for partition keys that I'm working on"

Can you please share more light what kids of characteristics partition columns have which we would like to leverage? And why cant those characteristics be some statistics based concepts?

Reply:
For example if it's a single partition (including subpartitions), we can replace things group by that column with constants (and subsequent optimizations can leverage that to say make it a global agg or something). You can't do that reliably just using stats. So yes you can do some things opportunistic things that may or may not work using stats but for doing stricter things (like replacing variable with constants) you will stronger guarantees that partition key constraints can give you.

Reply:
> > "As for column statistics - I don't think they are reliable"
> 
> Can you please share more light on it? We rely on them for fragment result caching and partition pruning.

Partition pruning is based on stats? that can't be. It should be using some kind of constraint solving.

Reply:
Check: PARTITION_STATISTICS_BASED_OPTIMIZATION_ENABLED

Reply:
For this case it sounds like you can use the statsEstimates and check isConfident(), which for cbo stats will only return true for table scan stats directly from the metastore, values nodes or project nodes that don't do changes to the columns.  It will also return true for all HBO stats.

Reply:
> For this case it sounds like you can use the statsEstimates and check isConfident(), which for cbo stats will only return true for table scan stats directly from the metastore, values nodes or project nodes that don't do changes to the columns. It will also return true for all HBO stats.

I would still vote for this API (maybe that statsbased one can be the implementation). Right now a lot of our table/column properties are not very accessible and that maybe preventing optimizations. In particular I'm really interested in partition keys for joins/aggs/window.

Reply:
> > For this case it sounds like you can use the statsEstimates and check isConfident(), which for cbo stats will only return true for table scan stats directly from the metastore, values nodes or project nodes that don't do changes to the columns. It will also return true for all HBO stats.
> 
> I would still vote for this API (maybe that statsbased one can be the implementation). Right now a lot of our table/column properties are not very accessible and that maybe preventing optimizations. In particular I'm really interested in partition keys for joins/aggs/window.

Basically I believe this is a fundamental/defined property of the column and should not be stats-based.

Reply:
We can look at that question too separately, but in this case it really seems you are just looking for low number of distinct values, and we already have a way to check that that will be more generic.

Reply:
> We can look at that question too separately, but in this case it really seems you are just looking for low number of distinct values, and we already have a way to check that that will be more generic.

I wanted to do for both because I'm going to have a different optimization for group by partition keys and I don't want this rule to trigger for that.

Reply:
what is the other optimization and what makes partition keys special there? Wouldn't it also be due to the low cardinality? So why not make it generic for all low cardinality columns?

Reply:
> what is the other optimization and what makes partition keys special there? Wouldn't it also be due to the low cardinality? So why not make it generic for all low cardinality columns?

For example - we can do something like "bucketed" execution for partition key aggregations where each file/split/page have the same value for the group by keys - so instead of hashing and aggregating row by row using input function, we can just give the entire array of values to the aggregation input functions (extending the api) and let it aggregate.

Reply:
There are two orthogonal issues I see:
- should column handle expose "isPartitionKey": I agree with sreeni that this is a fairly universal concept so IMHO it's ok to have that function here
- is partitioning information required for correctness or it's simply a proxy for "is low cardinality". For the latter we can create another wrapper function that for the time being only checks for whether the column is partitioned but in the future can look at HBO etc

Reply:
Discussed offline with @kaikalur that we can look at making partition key information accessible to the optimizer when it's actually needed (e.g. when the physical layout of the table is relevant), but in this case, what we care about is low distinctness, so keeping this more generic and looking at the distinct value stats (with isConfident() check) is a better approach here.

Reply:
> Discussed offline with @kaikalur that we can look at making partition key information accessible to the optimizer when it's actually needed (e.g. when the physical layout of the table is relevant), but in this case, what we care about is low distinctness, so keeping this more generic and looking at the distinct value stats (with isConfident() check) is a better approach here.

Yeah I started implementing a special optimization for partition keys and then noticed this issue so I paused that and wanted to sneak this by lol. Oh well - maybe I will just abandon this PR. let me think more

Reply:
Redid it. Please take a look

Reply:
But see how bad it is - bunch of constructors crashing all over the place. This is what I dislike about this stats frameworjk - it's not following Law of Demeter properly. These should be encapsulated better. I think I will go back to my partition column idea.